### PR TITLE
Update Flow to v0.146.0

### PR DESCRIPTION
**Pull request by wbinnssmith**
This updates Flow to v0.146.0 [0] , which invalidates refinements (particularly object properties and variables declared with `let` passed to functions) much more frequently. This required introducing additional invariants, `nullthrows`, or temporary locals, depending on the number of violations in a given scope.

This is probably a good motivation to use `const` to retain refinements of locals when Flow thinks they could be modified through function calls.

[0] https://github.com/facebook/flow/releases/tag/v0.146.0

### CODE REVIEW

File: packages/core/core/src/serializer.js [hunk 1]
```diff
@@ -119,9 +119,10 @@ function mapObject(object: any, fn: (val: any) => any, preOrder = false): any {
         if (result instanceof Map) {
           result.set(key, newValue);
         } else if (result instanceof Set) {
+          let _result = result; // For Flow
```
Comment (wbinnssmith) [inline packages/core/core/src/serializer.js:122]: This feels like a Flow bug? Why does introducing a new let binding resolve this?

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>React HackerNews 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.02m | +2.02m ⚠️ |
| Cached | 2.80s | +2.80s ⚠️ |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 14.70s | -116.00ms |
| Cached | 542.00ms | -25.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 580.60kb | +0.00b | 115.00ms | +12.00ms ⚠️ |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/292276496128541197)**

**Review by joeyslater — APPROVED** [review]

**Comment by mischnic** [comment]
> This is probably a good motivation to use const to retain refinements of locals when Flow thinks they could be modified through function calls.

But AFAICT, none of the cases where you needed to introduce another variable could be simplified with const, right?

**Comment by wbinnssmith** [comment]
> But AFAICT, none of the cases where you needed to introduce another variable could be simplified with const, right?

Yeah, none of these cases required it.

File: packages/core/core/src/BundlerRunner.js [hunk 2]
```diff
@@ -89,14 +89,20 @@ export default class BundlerRunner {
       cacheKey = await this.getCacheKey(graph, configResult);
       let cachedBundleGraphBuffer;
       try {
-        cachedBundleGraphBuffer = await this.options.cache.getBlob(cacheKey);
+        cachedBundleGraphBuffer = await this.options.cache.getBlob<Buffer>(
+          cacheKey,
+        );
       } catch {
         // Cache miss
```
Comment (mischnic) [inline packages/core/core/src/BundlerRunner.js:96]: Can `getBlob` actually nullish? I just does `return readFile()`?

File: packages/core/core/src/Transformation.js [hunk 3]
```diff
@@ -825,10 +826,12 @@ export default class Transformation {
 
     // Create generate function that can be called later
     pipeline.generate = (input: UncommittedAsset): Promise<GenerateOutput> => {
-      if (transformer.generate && input.ast) {
-        let generated = transformer.generate({
+      let generate = transformer.generate?.bind(transformer);
+      let ast = input.ast;
+      if (generate && ast) {
+        let generated = generate({
           asset: new Asset(input),
```
Comment (mischnic) [inline packages/core/core/src/Transformation.js:833]:
```suggestion
      let ast = input.ast;
      let asset = new Asset(input);
      if (transformer.generate && ast) {
        let generated = transformer.generate({
          asset,
```

File: packages/core/core/src/Validation.js [hunk 4]
```diff
@@ -90,8 +90,9 @@ export default class Validation {
               await Promise.all(
                 assets.map(async asset => {
                   let config = null;
-                  if (plugin.getConfig) {
-                    config = await plugin.getConfig({
+                  let getConfig = plugin.getConfig?.bind(plugin);
+                  if (getConfig) {
+                    config = await getConfig({
```
Comment (mischnic) [inline packages/core/core/src/Validation.js:95]: This can also be worked around by factoring `let publicAsset = new Asset()` out before the if statement

**Comment by wbinnssmith** [comment]
I'm going to land this but we need to address the TODO here: https://github.com/parcel-bundler/parcel/pull/5971/files#diff-5164e4e7a234bd85560a2fc9f18733e38d3de8892f324deeecec1417905e1ac3R582

This was uncovered when changing getBlog to not be polymorphic. I'll open an issue for this.

**Pull request merged**